Add example for dynamic messages

luca-della-vedova opened the pull request:
Split off https://github.com/ros2-rust/ros2_rust/pull/492 now that examples are in a different repo.
Will need the upstream merged and re-released with a new version for it to actually work (otherwise I can hardcode a branch / path if we want CI to be tested)

luca-della-vedova committed:
Add example for dynamic messages

Signed-off-by: Luca Della Vedova <lucadv@intrinsic.ai>

esteve commented on rclrs/dynamic_pub_sub/Cargo.toml:
```diff
@@ -0,0 +1,11 @@
+
+[dependencies]
+# TODO(luca) change this to the correct version once dynamic message support is released on crates.io
+rclrs = { version = "0.4", features = ["dyn_msg"] }
```
Now that `dyn_msg` is no longer a flag in `rclrs`, could you remove it here?

luca-della-vedova replied:
I removed it in https://github.com/ros2-rust/examples/pull/7/commits/7f8ac88cf63d118646d7d62bc786238ca5676c54 but note that as the TODO says this will only be merge-able after the upstream PR is merged and we have a re-release.
To make it compile-able I temporarily set it as a git dependency on my branch, otherwise this can't really be tested.

esteve reviewed (changes requested):
@luca-della-vedova thanks! Does this work for you without a `package.xml` file? Perhaps we don't need it for the examples, anyway.

luca-della-vedova committed:
Remove dyn_msg feature, change to git dependency

Signed-off-by: Luca Della Vedova <lucadv@intrinsic.ai>

luca-della-vedova committed:
Add dynamic publisher, cleanup Cargo.toml

Signed-off-by: Luca Della Vedova <lucadv@intrinsic.ai>

luca-della-vedova committed:
Add package.xml

Signed-off-by: Luca Della Vedova <lucadv@intrinsic.ai>

luca-della-vedova commented:
I haven't really put a lot of work into this PR and was leaving it as a "cleanup after the upstream is merged" but after your comment I realized that this could really be a good test platform to make sure the upstream PR is working as intended, so I cleaned this up and added functionality.

> @luca-della-vedova thanks! Does this work for you without a `package.xml` file? Perhaps we don't need it for the examples, anyway.

It's an interesting question, "it builds" but upon further inspection it's more complicated than that.
Technically, it builds because it doesn't strictly depend on any message to compile, since those are only loaded dynamically at runtime.
Furthermore, since it doesn't need rust bindings for the messages but just the dynamic library to be available we don't really need the `colcon-ros-cargo` machinery to make sure `cargo` can find the message at build time, but we just need the C++ library to be available in the `ament_prefix`.
Still, there is a mechanism for this kind of dependency, and it's called `exec_depend` in `package.xml` so indeed you are right and we should have a `package.xml` to be fully safe. I added it in https://github.com/ros2-rust/examples/pull/7/commits/14d032e93791a01b3a59ff1e891e161affc88322.

I also noticed that there was no dynamic_publisher examples, so I added that in https://github.com/ros2-rust/examples/pull/7/commits/5816dbdf3d3a350caa08303727ede092bddc9f24. I also changed the message type into the standard `example_interfaces/msg/String`, which is the same used in the `minimal_pub_sub` example.
Finally, I noticed that building it through `colcon-cargo` rather than `colcon-ros-cargo` I couldn't `ros2 run` it, unsure why at this moment, I'll investigate in `colcon-cargo` itself.

This means that, after all these changes, you can finally use this PR to test the upstream PR and even mix and match normal and dynamic pub sub! For example:

```bash
# You can of course try this
ros2 run examples_rclrs_dynamic_pub_sub dynamic_publisher
ros2 run examples_rclrs_dynamic_pub_sub dynamic_subscriber

# But also
ros2 run examples_rclrs_minimal_pub_sub minimal_publisher
ros2 run examples_rclrs_dynamic_pub_sub dynamic_subscriber

# And also!
ros2 run examples_rclrs_dynamic_pub_sub dynamic_publisher
ros2 run examples_rclrs_minimal_pub_sub minimal_subscriber
```

esteve commented:
@luca-della-vedova thanks for all the work! Now that https://github.com/ros2-rust/ros2_rust/pull/492 is merged and the new release is out, I think this could be merged. Are there any other changes you'd like to make to this PR?

esteve commented on rclrs/dynamic_pub_sub/src/dynamic_publisher.rs:
```diff
@@ -0,0 +1,38 @@
+
+    let node = executor.create_node("minimal_publisher")?;
+
+    let message_type = MessageTypeName {
```
@luca-della-vedova could you replace this with `"example_interfaces/msg/String".try_into()?` ? Or you wrote it like this so that users can see a different way for setting the message type?

luca-della-vedova replied:
Both ways work, the string is a bit less verbose but introduces a result that users have to take care of, while using the struct is an infallible way so users have one less error source.
But I'm really very neutral about it, if you think it should be replaced happy to
